sklearn/utils/metaestimators.py

```diff
@@ -32,41 +37,52 @@ def __get__(self, obj, type=None):
         if obj is not None:
             # delegate only on instances, not the classes.
             # this is to allow access to the docstrings.
-            self.get_attribute(obj)
+            if self.fn_check_obj is not None:
+                self.fn_check_obj(obj)
+            delegate = attrgetter(self.delegate_name)(obj)
+            getattr(delegate, self.attribute_name)
+
         # lambda, but not partial, allows help() to work with update_wrapper
         out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)
         # update the docstring of the returned function
         update_wrapper(out, self.fn)
         return out
 
 
-def if_delegate_has_method(delegate):
+def if_delegate_has_method(delegate, fn_check_obj=None):
     """Create a decorator for methods that are delegated to a sub-estimator
 
     This enables ducktyping by hasattr returning True according to the
     sub-estimator.
 
-    >>> from sklearn.utils.metaestimators import if_delegate_has_method
-    >>>
-    >>>
-    >>> class MetaEst(object):
-    ...     def __init__(self, sub_est):
-    ...         self.sub_est = sub_est
-    ...
-    ...     @if_delegate_has_method(delegate='sub_est')
-    ...     def predict(self, X):
-    ...         return self.sub_est.predict(X)
-    ...
-    >>> class HasPredict(object):
-    ...     def predict(self, X):
-    ...         return X.sum(axis=1)
-    ...
-    >>> class HasNoPredict(object):
-    ...     pass
-    ...
-    >>> hasattr(MetaEst(HasPredict()), 'predict')
-    True
-    >>> hasattr(MetaEst(HasNoPredict()), 'predict')
-    False
+    Parameters
+    ----------
+    delegate : string
+        Name of the sub-estimator that can be accessed as an attribute of the
+        base object.
+
+    fn_check_obj : function
+        fn_check_obj(obj) with obj as the metaestimator instance.
+        The function will be executed before ducktyping. Typically used for
+        checking if the estimator has been fitted.
+
```

Comment:
I would leave the doctests as they were or add them as tests to sklearn/utils/tests/test_metaestimators.py.

You could also add additional tests in the same file as you did at some point of this PR.